src/analysis/processing/qgsalgorithmzonalstatistics.cpp

```diff
@@ -89,14 +89,14 @@ void QgsZonalStatisticsAlgorithm::initAlgorithm( const QVariantMap & )
   addParameter( new QgsProcessingParameterRasterLayer( QStringLiteral( "INPUT_RASTER" ), QObject::tr( "Raster layer" ) ) );
   addParameter( new QgsProcessingParameterBand( QStringLiteral( "RASTER_BAND" ),
                 QObject::tr( "Raster band" ), 1, QStringLiteral( "INPUT_RASTER" ) ) );
-  addParameter( new QgsProcessingParameterVectorLayer( QStringLiteral( "INPUT_VECTOR" ), QObject::tr( "Vector layer containing zones" ),
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT_VECTOR" ), QObject::tr( "Feature source containing zones" ),
                 QList< int >() << QgsProcessing::TypeVectorPolygon ) );
   addParameter( new QgsProcessingParameterString( QStringLiteral( "COLUMN_PREFIX" ), QObject::tr( "Output column prefix" ), QStringLiteral( "_" ) ) );
 
   addParameter( new QgsProcessingParameterEnum( QStringLiteral( "STATISTICS" ), QObject::tr( "Statistics to calculate" ),
                 statChoices, true, QVariantList() << 0 << 1 << 2 ) );
 
-  addOutput( new QgsProcessingOutputVectorLayer( QStringLiteral( "INPUT_VECTOR" ), QObject::tr( "Zonal statistics" ), QgsProcessing::TypeVectorPolygon ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Zonal statistics" ), QgsProcessing::TypeVectorPolygon, QVariant(), false, true, true ) );
```

Comment:
Unfortunately we can't make this change here -- it would break all existing processing models and scripts which use this algorithm.

That said -- I do think this is a very good change. What we'd need to do is
1. split the algorithm into two (with a common base class).
2. the existing native:zonalstatistics algorithm remains unchanged, with the exception that it now returns the FlagDeprecated flag to hide it from users
3. the new subclass does your new logic and creates a new sink, instead of editing in place. (and this one can remove the FlagNoThreading flag). It has a different id, so as not to clash with the existing deprecated algorithm.

Reply:
Fair points. I didn't like the approach that much anyway, completely different one coming up.